src/quota.rs

```diff
@@ -0,0 +1,201 @@
+        let folders = get_watched_folders(context).await;
+        let unique_quota_roots = get_unique_quota_roots_and_usage(folders, imap).await?;
+
+        // build and send report message
+        let mut msg = Message::new(Viewtype::Text);
+        msg.text = Some(generate_report_message(&unique_quota_roots, context).await?);
+        add_device_msg(context, None, Some(&mut msg)).await?;
+    } else {
+        warn!(
+            context,
+            "the email server does not support the quota extention"
+        );
+        let mut msg = Message::new(Viewtype::Text);
+        msg.text = Some(quota_not_supported(context).await); // todo stock string?
+        add_device_msg_with_importance(context, None, Some(&mut msg), true).await?;
+    }
+
+    Ok(())
+}
+
+async fn generate_report_message(
+    unique_quota_roots: &[(String, Vec<QuotaResource>)],
+    context: &Context,
+) -> Result<String> {
+    let mut message: String = "".to_owned();
+
+    let storage_stock_string = quota_resource_storage(context).await;
+    let messages_stock_string = quota_resource_messages(context).await;
+    for (name, quota_resources) in unique_quota_roots {
+        message.push_str(&format!("{}:\n", &name));
+        use async_imap::types::QuotaResourceName::*;
+        for resource in quota_resources {
+            message.push_str(&match &resource.name {
+                Atom(name) => {
+                    format!("[{}/{}] {}\n", resource.usage, resource.limit, name)
+                }
+                Message => {
+                    format!(
+                        "{}/{} {}\n",
+                        resource.usage, resource.limit, messages_stock_string
+                    )
+                }
+                Storage => {
+                    let used = (resource.usage * 1024)
+                        .file_size(file_size_opts::BINARY)
+                        .map_err(|err| anyhow!("{}", err))?;
+                    let limit = (resource.limit * 1024)
+                        .file_size(file_size_opts::BINARY)
+                        .map_err(|err| anyhow!("{}", err))?;
+                    format!("{}/{} {}\n", used, limit, storage_stock_string)
+                }
+            });
+        }
+    }
+    Ok(message)
+}
+
+async fn get_unique_quota_roots_and_usage(
+    folders: Vec<String>,
+    imap: &mut Imap,
+) -> Result<Vec<(String, Vec<QuotaResource>)>> {
+    // IDEA: OPTIMIZATION:
+    // the unique quota roots of get_unique_quota_roots_and_usage could be cached and then the server could be asked for the quotas of those.
+    // this would reduce incoming traffic and for most cases also outgoing traffic, because most email server share quota roots across folders/mailboxes.
+    // Again at the cost of increasing code complexity and the question for how long it should be cached
+    // IDEA2: maybe the provider db could also contain quota root names?
+    let mut unique_quota_roots: Vec<(String, Vec<QuotaResource>)> = Vec::new();
```

Comment:
Looks like what you want is a `HashMap` or `indexmap::IndexMap` (we already use `indexmap` anyway) instead of using `find_position` each time to go through vectors. Maps have a nice entry API that allow modifying the entry in-place without `get_mut(position)`.